### ISSUE TITLE
0.0.21 regression - `invalid-assignment` false positive when assigning type with invariant generic to a `TypedDict` key when the generic should instead be inferred as a wider type

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#2986]
### Summary

```py
from typing import TypedDict

class Foo:
    asdf: list[int | str]

class Bar(TypedDict):
    asdf: list[int | str]


def fn(foo: Foo, bar: Bar):
    bar["asdf"] = list([1]) # error: Invalid assignment to key "asdf" with declared type `list[int | str]` on TypedDict `Bar`: value of type `list[int]` (invalid-assignment)
    foo.asdf = list([1]) # no error, correctly inferred as list[int | str]
```

https://play.ty.dev/2e780e02-941e-405b-b65e-6c61a81857cf

this bug seems to have been introduced in v0.0.21

### Version

0.0.21

**Comment by carljm** [comment]
Thanks for the report! This isn't a regression in bidirectional inference, it's an existing gap that shows up now because we stopped unioning in `Unknown` to the inferred type of invariant containers (#1240).

**Comment by carljm** [comment]
@ibraheemdev Correct me if I'm wrong, but I think that bidirectional inference issues like this one are independent of #1473, in that #1473 will expand the locations we look to for sources of type context, but improving the way we interpret that type context is somewhat orthogonal?

**Comment by ibraheemdev** [comment]
Yeah, there are a couple of known places where we don't currently infer with type context, especially around dunder calls. I opened https://github.com/astral-sh/ruff/pull/23800 to fix this one.
